src/vmemcache.c

```diff
@@ -60,6 +78,92 @@ vmemcache_newU(const char *dir, size_t max_size, size_t fragment_size,
 	LOG(3, "dir %s max_size %zu fragment_size %zu replacement_policy %d",
 		dir, max_size, fragment_size, replacement_policy);
 
+	if (max_size < VMEMCACHE_MIN_POOL) {
+		ERR("size %zu smaller than %zu", max_size, VMEMCACHE_MIN_POOL);
+		errno = EINVAL;
+		return NULL;
+	}
+
+	if (fragment_size < VMEMCACHE_MIN_FRAG) {
+		ERR("fragment size %zu smaller than %zu bytes",
+			fragment_size, VMEMCACHE_MIN_FRAG);
+		errno = EINVAL;
+		return NULL;
+	}
+
+	if (fragment_size > max_size) {
+		ERR(
+			"fragment size %zu larger than maximum file size: %zu bytes",
+			fragment_size, max_size);
+		errno = EINVAL;
+		return NULL;
+	}
+
+	VMEMcache *cache = Zalloc(sizeof(VMEMcache));
+	if (cache == NULL) {
+		ERR("!Zalloc");
+		return NULL;
+	}
+
+	enum file_type type = util_file_get_type(dir);
+	if (type == OTHER_ERROR) {
+		LOG(1, "checking file type failed");
+		goto error_free_cache;
+	}
+
+	if (type == TYPE_DEVDAX) {
+		const char *devdax = dir;
+		ssize_t size = util_file_get_size(devdax);
+		if (size < 0) {
+			LOG(1, "cannot determine file length \"%s\"", devdax);
+			goto error_free_cache;
+		}
+
+		if ((cache->addr = util_file_map_whole(devdax)) == NULL) {
+			LOG(1, "mapping of whole DAX device failed");
+			goto error_free_cache;
+		}
+
+		cache->size = (size_t)size;
+	} else {
+		/* silently enforce multiple of mapping alignment */
+		cache->size = roundup(max_size, Mmap_align);
+
+		if ((cache->addr = util_map_tmpfile(dir, cache->size,
```

Comment:
add:
`XXX: file should be mapped on-demand during allocation, up to cache->size` or something to that effect. Just so that we don't forget.

Reply:
Done